RFC-3

autoencoder opened the pull request:
Putting up a draft for discussion. Don't have to merge (yet.)
- Is the glossary correct?
- Should we move the glossary to a separate document?

bigs committed:
Sketch out server side RPC for transactor

vyzo committed:
scripts/ubuntu-setup.sh for setting up on ubuntu 14.0.4LTS

yusefnapora committed:
add insert / update rpc endpoints

bigs committed:
Add static method to launch a TransactorService

bigs committed:
Simplify and streamline install script.

Should be ready for non-interactive runs.

bigs committed:
Revert datastore type change

parkan committed:
Update README.md

yusefnapora committed:
scala app to start gRPC TransactorService

yusefnapora committed:
no sleep for the wicked

vyzo committed:
transactor/README: add instructions for generating keystores for SSL

vyzo committed:
minor: capitalize howto

vyzo committed:
transactor/README: fix argument in keystore howto

vyzo committed:
transactor/README: polish the ssl howto comments

autoencoder committed:
Reorganize sections, media identifiers, extra terminology

yusefnapora committed:
Alter server-side API to just fetch refs

yusefnapora committed:
Refactor Tx service to use copycat client

bigs committed:
Fix implicit arg & use approp hash serialization

yusefnapora committed:
fix RpcService app to use new TransactorService args

yusefnapora committed:
cbor serializer for JournalInsert command

parkan committed:
Merge pull request #107 from mediachain/vyzo-transactor-scripts

Script for basic setup of ubuntu 14.0.4LTS instances

parkan committed:
Merge pull request #108 from mediachain/vyzo-keystore-openssl

Instructions for generating keystores for copycat  with OpenSSL

yusefnapora committed:
add TypeSerializer for JournalUpdate command

yusefnapora committed:
special-case deserializing References, since they lack a type field

yusefnapora committed:
add TypeSerializer for JournalLookup query

yusefnapora committed:
better names and status codes for TransactorService ops

yusefnapora committed:
spec for copycat cbor serializers

parkan commented on rfc/mediachain-rfc-3.md:
```diff
@@ -14,15 +14,22 @@ We will also outline a roadmap toward improved, `Future` Indexer generations. Fu
+- `Artefact` - Media work tracked by Mediachain. Identified in the blockchain by its latest `Canonical ID`.
+- `Entity` - Persona or organization who may associate with `Artefacts`. Identified in the blockchain by its latest `Canonical ID`.
+- `Canonical` -  Serialized blockchain block, stored in IPFS and identified by a `Canonical ID`.
+- `Canonical ID` - IPFS ID of a serialized blockchain block. In addition to identifying blockchain blocks, also used by blockchain to identify `Entities` and `Artifacts`. The current `Canonical IDs` of an `Entity` or `Artefact` can change over time, when its metadata is updated in the blockchain or when it is merged with another `Entity` or `Artefact`.
```
I don't think the Canonical and Canonical ID definitions are right--we've usually spoken of a "canonical" as the UUID-like identifier given to our "idea" of a work of art to which all data about it and its "instances" (copies, appearances in media, etc) refer to, and the "canonical id" as that identifier. In the current version, I **think** the correct analogy is the `Artefact`/`Entity` multihash (which is pointed to by `ref` in an `ArtefactCreationCell` etc) but we should clear this up.

parkan replied:
the way that you're using this elsewhere it seems like maybe you want just "multihash" or "IPFS address"

parkan replied:
(so we should sort out the terminology around this, particularly the ID)

autoencoder replied:
It'll be great to clarify how "idea" versus "instance" are each tracked in the system.

While we're at it, worth keeping in mind exactly how we'll define "idea" -- e.g. what's the particular definition of a sameness. The supervised indexer models will support being separately trained on multiple, potentially conflicting, definitions of sameness. At a minimum, 2 types - duplicate similarity and search query relevance similarity. But there can be multiple conflicting sub-types to choose from, for each of those as well.

At some point, also worth thinking about whether the chain or end-user applications might support tracking multiple conflicting sameness definitions, possibly for applying to different purposes.

parkan commented on rfc/mediachain-rfc-3.md:
```diff
@@ -14,15 +14,22 @@ We will also outline a roadmap toward improved, `Future` Indexer generations. Fu
 - `Clients` - Embedded API or external nodes, which are the interface for end-user nodes to receive a stream of updates to the `Journal` from `Transactor` nodes, and for communicating writes to the `Transactors`. They also conduct reads and writes to the `Datastore`.
 - `Transactors` - Core of Mediachain. Accepts writes to the `Journal` from `Clients`, serves `Clients` with views of the `Journal`, and records transactions on their behalf. (Referred to as `Peer Nodes` in RFC-2.)
-- `Media Identifier` - One of several types of strings that can be used to indicate the identity of a media work. Most Indexer subsystems will resolve this identifier into a canonical ID. See more in the `Media Identifiers` section.
+- `Media Identifier` - String used to identify a media work. The string's prefix determines which, of several possible identifier types, each string is. See the `Media Identifiers` section in this RFC.
```
Sorry, I guess I'm reading this out of order -- these are what we'd call "canonical ids" (at least the IPFS style `Qm...` one)

autoencoder replied:
So the idea with the `Media Identifiers` is to have a name for all types of image identifiers which can be accepted by the Indexer APIs. In other words, `canonical IDs` are a subset of the possible identifier types the Indexer API can accept. Also currently supported is querying by image content e.g. data URIs of small images, or could even extend it to querying by image descriptors / vectors that represent abstract visual concepts instead of a particular image's content - as is sometimes done in some image QA systems.

parkan replied:
Yeah, totally agreed that all of these are valid "identifiers" but I think we should pick something as first class, which is the canonical/IPFS address in this case. Conversely, I don't think IPFS addresses of other IPFS objects should be considered "canonical ids". So all canonical ids are IPFS addresses, but not all IPFS addresses are canonical ids.

parkan commented on rfc/mediachain-rfc-3.md:
```diff
@@ -14,15 +14,22 @@ We will also outline a roadmap toward improved, `Future` Indexer generations. Fu
+- `Media Identifier` - String used to identify a media work. The string's prefix determines which, of several possible identifier types, each string is. See the `Media Identifiers` section in this RFC.
+- `Artefact` - Media work tracked by Mediachain. Identified in the blockchain by its latest `Canonical ID`.
+- `Entity` - Persona or organization who may associate with `Artefacts`. Identified in the blockchain by its latest `Canonical ID`.
+- `Canonical` -  Serialized blockchain block, stored in IPFS and identified by a `Canonical ID`.
```
I think this reflects some confusion about the design: conceptually, there are individual chains for each "Canonical" (aka Artefact/Entity), and a Journal with two parts: a blockchain containing blocks of (~1000 each) updates of the Chain Heads of these chains, and an index that points to the current chain head for each one. This is a bit of an unorthodox design and keeps tripping people up, we can do a call tomorrow with @vyzo (please correct me if my explanation here is off, too) to clear this up.

parkan commented on rfc/mediachain-rfc-3.md:
```diff
@@ -14,15 +14,22 @@ We will also outline a roadmap toward improved, `Future` Indexer generations. Fu
+- `Entity` - Persona or organization who may associate with `Artefacts`. Identified in the blockchain by its latest `Canonical ID`.
+- `Canonical` -  Serialized blockchain block, stored in IPFS and identified by a `Canonical ID`.
+- `Canonical ID` - IPFS ID of a serialized blockchain block. In addition to identifying blockchain blocks, also used by blockchain to identify `Entities` and `Artifacts`. The current `Canonical IDs` of an `Entity` or `Artefact` can change over time, when its metadata is updated in the blockchain or when it is merged with another `Entity` or `Artefact`.
+- `Chain Head` - `Canonical ID` of latest blockchain block committed to the `Journal`. Multiple `Chain Heads` are permitted to exist during temporary blockchain forks.
```
This would be something like "the `head` field of the latest ChainEntry commited to the Journal for a given Canonical, referring to the latest ChainCell for that Canonical"

parkan commented:
Left a bunch of specific comments, but the main takeaways are:
- the many chains + journal blockchain + journal index design isn't super clear from existing docs, we should 1) sync up the team to make sure we are on the same page now 2) create a doc (maybe not RFC format, something visual?) explaining this core design in simple terms
- need to clear up meaning and nature of IDs
- the glossary should probably be separate from the RFC, see also https://github.com/mediachain/mediachain/issues/29

vyzo committed:
copycat serializers: refactor common code for reading and writing bytes

vyzo committed:
copycat serializers: simplify code for JournalBlock/Insert deserializer using fromCborBytes

vyzo committed:
Merge pull request #111 from mediachain/yn-command-serializers

Serialize copycat commands as CBOR bytes

vyzo committed:
scripts/ubuntu-setup: fixes for oracle jdk license

vyzo committed:
copycat/StateMachine: reduce blocksize to 512 in order to alleviate dynamo provisioning requirements

With 512 entries/block there is no need for chunking, which makes expensive proviosioning for the chunk table redundant.

vyzo committed:
copycat client: support cluster connections

vyzo committed:
Merge branch 'vyzo-copycat-cluster-client' into vyzo-copycat-misc

parkan committed:
Merge pull request #113 from mediachain/vyzo-copycat-misc

copycat: reduce blocksize, add cluster connect to client

vyzo committed:
Merge pull request #112 from mediachain/vyzo-script-fixes

scripts/ubuntu-setup fixes

vyzo committed:
transactor: signal JournalDuplicateError on duplicate inserts

vyzo committed:
Merge pull request #114 from mediachain/vyzo-copycat-journal-insert

transactor: signal JournalDuplicateError on duplicate inserts

yusefnapora committed:
bump grpc version

denisnazarov committed:
Add some links to README

yusefnapora committed:
Merge origin/master into bigs-yn-transactor-rpc

vyzo committed:
PersistentDatastore: log queue size on backoffs

vyzo committed:
JournalServer: add interactive control for shutdown/cluster leave

vyzo committed:
Transactor protocol: give interface methods better names

vyzo committed:
TransactorService: Add max record size check to insert/update

vyzo committed:
cosmetics

vyzo committed:
more cosmetics: errant line in Services

vyzo committed:
transactor RpcService: support cluster connect

vyzo committed:
transactor RpcService: kind programs println their error messages

parkan committed:
Merge pull request #110 from mediachain/bigs-yn-transactor-rpc

RPC client interface for transactor

vyzo committed:
JournalServer: more idiomatic arg matching

vyzo committed:
Merge pull request #115 from mediachain/vyzo-transactor-misc

Tweaks in transactor server

vyzo committed:
protocol: add metaSource to Record data model

vyzo committed:
update depedents for Record#metaSource

vyzo committed:
rfc-1: add metaSource to data model

vyzo committed:
Merge pull request #118 from mediachain/vyzo-meta-source

Make original metadata reference in records a first-class citizen

autoencoder committed:
Merge branch 'autoencoder-drafts' of https://github.com/mediachain/mediachain into autoencoder-drafts

autoencoder committed:
Reorganize RFC-3.
